src/lib/uptime_service/uptime_service.ml

```diff
@@ -0,0 +1,431 @@
+                  ; ("http_code", `Int (Cohttp.Code.code_of_status status))
+                  ; ( "http_error"
+                    , `String (Cohttp.Code.string_of_status status) ) ] )
+  with
+  | Ok () ->
+      ()
+  | Error exn ->
+      [%log error]
+        "Error when sending block with state hash $state_hash to uptime \
+         server at URL $url"
+        ~metadata:
+          [ ("state_hash", State_hash.to_yojson state_hash)
+          ; ("url", `String (Uri.to_string url))
+          ; ("error", `String (Exn.to_string exn)) ]
+
+let block_base64_of_breadcrumb ~logger breadcrumb =
+  let external_transition = external_transition_of_breadcrumb breadcrumb in
+  let block_len =
+    External_transition.Stable.Latest.bin_write_t block_buf ~pos:0
+      external_transition
+  in
+  let block_string = String.init block_len ~f:(fun ndx -> block_buf.{ndx}) in
+  let block_base64_result = Base64.encode block_string in
+  match block_base64_result with
+  | Ok block_base64 ->
+      Some block_base64
+  | Error (`Msg err) ->
+      [%log error] "Could not Base64-encode block with state hash $state_hash"
+        ~metadata:
+          [ ( "state_hash"
+            , State_hash.to_yojson
+              @@ External_transition.state_hash external_transition )
+          ; ("error", `String err) ] ;
+      None
+
+let send_produced_block_at ~logger ~url ~transition_frontier ~peer_id
+    ~(submitter_keypair : Keypair.t) tm =
+  (* give block production some time *)
+  let%bind () = at (Time.add tm @@ Time.Span.of_min 2.0) in
+  match Block_producer.last_block_produced_opt () with
+  | None ->
+      [%log error] "State hash of last block produced unavailable"
+        ~metadata:
+          [ ( "expected_block_production_time"
+            , `String (Time.to_string_abs ~zone:Time.Zone.utc tm) ) ] ;
+      return ()
+  | Some state_hash -> (
+    match Broadcast_pipe.Reader.peek transition_frontier with
+    | None ->
+        [%log error]
+          "Transition frontier not available to obtain produced block with \
+           state hash $state_hash"
+          ~metadata:[("state_hash", State_hash.to_yojson state_hash)] ;
+        return ()
+    | Some tf -> (
+      match Transition_frontier.find tf state_hash with
+      | None ->
+          [%log error]
+            "Produced block with state hash $state_hash not found in \
+             transition frontier"
+            ~metadata:[("state_hash", State_hash.to_yojson state_hash)] ;
+          return ()
+      | Some breadcrumb -> (
+        match block_base64_of_breadcrumb ~logger breadcrumb with
+        | Some block_base64 ->
+            let block_data =
+              { block= block_base64
+              ; created_at= get_rfc3339_time ()
+              ; peer_id
+              ; snark_work= None }
+            in
+            send_uptime_data ~logger ~submitter_keypair ~url ~state_hash
+              block_data
+        | None ->
+            return () ) ) )
```

Comment:
no data to send in this case, but not an error?

Reply:
the error is logged in the call that produced the `None`.